design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,119 @@
+# Overview
+VM Cloning is the action of creating a new VM from an existing one that has the exact same setting. This includes
+VM disks, network definition, CPU topology and so on.
+
+## Motivation
+The ability to clone a VM is a generic virtualization operation which may be useful for several use-cases.
+
+For example, a system admin / user might want to base a new VM on an existing one. In such a case we can clone a VM and
+then change it according to our needs. In addition, a cluster admin might want to try different configurations
+(for example, to check whether other configuration provide better performance for his workload), in this case he can
+clone the existing VM, test it and decide whether to move the new settings.
+
+Naively copy/pasting a VM spec would not work in most cases therefore this serves as an automation for cloning safely.
+
+## Goals
+Have a mechanism to clone VMs. Original and cloned VM would have the same:
+* Disks
+* Network configuration
+* Volumes
+* CPU topology
+
+Also, provide an ability to fine-tune what's being cloned. For example, it is possible to clone or skip
+annotations / labels.  
+
+## Non Goals
+In first implementation iteration the following limitations exist:
+* Only offline cloning will be supported, meaning cloning a VM when it's stopped.
+* Cloned VM would be in the same namespace as the original one.
+* Host disks would not be supported.
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
+* The VM that needs to be cloned
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Whether to clone or skip labels / annotations etc.
```

Comment:
What is the default?
In general, this sounds risky to me. Some future annotation may not consider this cloning to exist and cause issues.

Reply:
The default IMO should be to copy all annotations / labels.
In the future we can exclude certain annotations that we know are problematic. Do you know of any current annotations that we should exclude?

Reply:
Annotation `kubevirt.io/storage-observed-api-version` & `kubevirt.io/latest-observed-api-version` comes to my mind.

Reply:
Perhaps we can solve this by having a test that expects X labels / annotations to be cloned. If we clone more than X it would fail. This way we can raise attention for added labels / annotations that are to be cloned.

In any case, I think that demanding a complete and explicit list of values to clone for every category (labels, annotations, volumes, etc) doesn't make sense - it would make cloning cumbersome and pretty useless. Furthermore, we can't demand the user to decide which of **our labels** should be cloned or not. If we're having hard time knowing it, how would the user know?

If managing our own labels / annotations is too difficult, maybe we can define several prefixes that would categorize them so it's easier to reason about their purpose. In one way or another - I think we should handle this - not the user.